### PR TITLE
Add input type validation to feed_ndarray in MXNet and PyTorch

### PR DESCRIPTION
**Pull request by JanuszL**
- adds type validation between DALI Tensor/TensorList and MXNet/PyTorch tensor
  inside feed_ndarray just in case anyone wants to use feed_ndarray directly
- it doesn't cover PaddlePaddle as feed_ndarray accepts a raw pointer to
  PaddlePaddle tensor and there is no API to check the type
- updates usage of raises, assert_raises in test_fw_iterators.py to use
  implementation from nose_utils

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

## Description
- [ ] **Bug fix** (*non-breaking change which fixes an issue*)
- [x] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR
- adds type validation between DALI Tensor/TensorList and MXNet/PyTorch tensor
  inside feed_ndarray just in case anyone wants to use feed_ndarray directly
- it doesn't cover PaddlePaddle as feed_ndarray accepts a raw pointer to
  PaddlePaddle tensor and there is no API to check the type
- updates usage of raises, assert_raises in test_fw_iterators.py to use
  implementation from nose_utils

#### Additional information
- Affected modules and functionalities:
  -  test_fw_iterators.py
  - plugin/mxnet.py
  - plugin/pytorch.py

- Key points relevant for the review:

### Tests
- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

File: dali/python/nvidia/dali/plugin/mxnet.py [hunk 1]
```diff
@@ -52,6 +52,15 @@ def feed_ndarray(dali_tensor, arr, cuda_stream = None):
                     In most cases, using the default internal user stream or stream 0
                     is expected.
     """
+    if isinstance(dali_tensor, (TensorListCPU, TensorListGPU)):
```
Comment (JanuszL) [inline dali/python/nvidia/dali/plugin/mxnet.py:55]: We cannot do such thing for PaddlePaddle as `feed_ndarray` there accepts raw pointer. We cannot accept tensor itself there as we cannot extract pointer itself to it without providing placement and type, so we would need to extend `feed_ndarray` signature and move allocation of data there as well (setting shape and placement does that). See [the API](https://github.com/PaddlePaddle/Paddle/blob/572bad8a66ea9650d753911af8383785ed671af9/paddle/fluid/pybind/pybind.cc#L638).

File: dali/test/python/test_fw_iterators.py [hunk 2]
```diff
@@ -1480,3 +1484,28 @@ def test_paddle_prepare_first_batch():
...
+
+def test_pytorch_feed_ndarray():
+    from nvidia.dali.plugin.pytorch import feed_ndarray
+    import torch
+
+    pipe = feed_ndarray_test_pipeline(batch_size=1, num_threads=1, device_id=0)
+    pipe.build()
+    out = pipe.run()[0]
+    torch_tensor = torch.empty((1), dtype=torch.int8, device = 'cpu')
+    assert_raises(AssertionError, feed_ndarray, out, torch_tensor, glob="Type of DALI Tensor/TensorList doesn't match Torch tensor type:*")
```
Comment (stiepan) [inline dali/test/python/test_fw_iterators.py:1511]: Just a nitpick, error message checking looks for any occurrence of the pattern, so stars are not necessary at the beginning and end.

**Review by stiepan — APPROVED** [review]

File: dali/test/python/test_fw_iterators.py (hunk 2 above)
Comment (JanuszL) [reply dali/test/python/test_fw_iterators.py]: Done

**Comment by JanuszL** [comment]
!build

File: dali/test/python/test_fw_iterators.py [hunk 3]
```diff
@@ -344,7 +345,8 @@ def check_mxnet_iterator_pass_reader_name(shards_num, pipes_number, batch_size,
 
     if batch_size > data_set_size // shards_num and last_batch_policy == LastBatchPolicy.DROP:
         assert_raises(AssertionError, MXNetIterator, pipes, [
-                      ("ids", MXNetIterator.DATA_TAG)], reader_name="Reader", last_batch_policy=last_batch_policy)
+                      ("ids", MXNetIterator.DATA_TAG)], reader_name="Reader", last_batch_policy=last_batch_policy,
+                      glob="It seems that there is no data in the pipeline. This may happen if `last_batch_policy` is set to PARTIAL and the requested batch size is greater than the shard size.")
```
Comment (mzient) [inline dali/test/python/test_fw_iterators.py:349]: I don't think we should put verbatim copies of the entire message here - just enough to make sure it's the error we expect - like:
```suggestion
                      glob="It seems that there is no data in the pipeline*last_batch_policy*")
```

File: dali/python/nvidia/dali/plugin/mxnet.py [hunk 4]
```diff
@@ -52,6 +52,15 @@ def feed_ndarray(dali_tensor, arr, cuda_stream = None):
...
+        dali_type = dali_tensor[0].dtype()
+    else:
+        dali_type = dali_tensor.dtype()
+    dali_type = np.dtype(dali_type)
+
+    assert dali_type == arr.dtype, ("Type of DALI Tensor/TensorList"
+            " doesn't match MXNet tensor type: {} vs {}".format(dali_type, np.dtype(arr.dtype)))
```
Comment (mzient) [inline dali/python/nvidia/dali/plugin/mxnet.py:62]:
```suggestion
    assert dali_type == arr.dtype, ("The element type of DALI Tensor/TensorList"
            " doesn't match the element type of the target MXNet NDArray {} vs {}".format(dali_type, np.dtype(arr.dtype)))
```

File: dali/python/nvidia/dali/plugin/pytorch.py [hunk 5]
```diff
@@ -51,6 +51,14 @@ def feed_ndarray(dali_tensor, arr, cuda_stream = None):
...
                     if we are copying to a tensor allocated with torch.zeros(...))
     """
+    if isinstance(dali_tensor, (TensorListCPU, TensorListGPU)):
+        dali_type = dali_tensor[0].dtype()
+    else:
+        dali_type = dali_tensor.dtype()
+    dali_type = np.dtype(dali_type)
+
+    assert to_torch_type[dali_type] == arr.dtype, ("Type of DALI Tensor/TensorList"
+            " doesn't match Torch tensor type: {} vs {}".format(to_torch_type[dali_type], arr.dtype))
```
Comment (mzient) [inline dali/python/nvidia/dali/plugin/pytorch.py:61]:
```suggestion
    assert to_torch_type[dali_type] == arr.dtype, ("The element type of DALI Tensor/TensorList"
            " doesn't match the element type of the target PyTorch Tensor: {} vs {}".format(to_torch_type[dali_type], arr.dtype))
```

File: dali/test/python/test_fw_iterators.py [hunk 6]
```diff
@@ -1480,3 +1484,28 @@ def test_paddle_prepare_first_batch():
...
+
+def test_mxnet_feed_ndarray():
+    from nvidia.dali.plugin.mxnet import feed_ndarray
+    import mxnet
+
+    pipe = feed_ndarray_test_pipeline(batch_size=1, num_threads=1, device_id=0)
+    pipe.build()
+    out = pipe.run()[0]
+    mxnet_tensor = mxnet.nd.empty([1], None, np.int8)
+    assert_raises(AssertionError, feed_ndarray, out, mxnet_tensor, glob="Type of DALI Tensor/TensorList doesn't match MXNet tensor type:")
```
Comment (mzient) [inline dali/test/python/test_fw_iterators.py:1500]: Update the pattern here if you update the message as suggested.

File: dali/python/nvidia/dali/plugin/pytorch.py (hunk 5 above)
Comment (JanuszL) [reply dali/python/nvidia/dali/plugin/pytorch.py]: Fixed

File: dali/python/nvidia/dali/plugin/mxnet.py (hunk 4 above)
Comment (JanuszL) [reply dali/python/nvidia/dali/plugin/mxnet.py]: Fixed

File: dali/test/python/test_fw_iterators.py (hunk 3 above)
Comment (JanuszL) [reply dali/test/python/test_fw_iterators.py]: Fixed

File: dali/test/python/test_fw_iterators.py (hunk 6 above)
Comment (JanuszL) [reply dali/test/python/test_fw_iterators.py]: Adjusted

**Comment by JanuszL** [comment]
!build

**Review by mzient — APPROVED** [review]

**Pull request merged**